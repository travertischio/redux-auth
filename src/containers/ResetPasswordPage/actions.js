/*
 *
 * ResetPasswordPage actions
 *
 */

import {
  RESET_PASSWORD_ACTION,
  RESET_PASSWORD_SUCCEED_ACTION,
  RESET_PASSWORD_FAILED_ACTION,
  DESTROY_PAGE_ACTION,
} from './constants';

export function resetPasswordAction(newPasswordPayload) {
  return {
    type: RESET_PASSWORD_ACTION,
    payload: newPasswordPayload,
  };
}

export function resetPasswordSucceedAction(response) {
  return {
    type: RESET_PASSWORD_SUCCEED_ACTION,
    payload: response,
  };
}

export function resetPasswordFailedAction(rejection) {
  return {
    type: RESET_PASSWORD_FAILED_ACTION,
    payload: rejection,
  };
}

export function destroyPageAction() {
  return {
    type: DESTROY_PAGE_ACTION,
  };
}
