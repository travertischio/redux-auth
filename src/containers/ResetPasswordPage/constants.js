/*
 *
 * ResetPasswordPage constants
 *
 */

export const RESET_PASSWORD_ACTION = 'app/ResetPasswordPage/RESET_PASSWORD';
export const RESET_PASSWORD_SUCCEED_ACTION = 'app/ResetPasswordPage/RESET_PASSWORD_SUCCEED';
export const RESET_PASSWORD_FAILED_ACTION = 'app/ResetPasswordPage/RESET_PASSWORD_FAILED';
export const DESTROY_PAGE_ACTION = 'app/ResetPasswordPage/DESTROY';
