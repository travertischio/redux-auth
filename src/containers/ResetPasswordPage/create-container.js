import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Helmet from 'react-helmet';
import { injectIntl } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import compose from 'recompose/compose';
import { AuthenticationContext } from '../AuthenticationProvider/hocs';
import selectResetPasswordPage from './selectors';
import messages from './messages';
import {
  resetPasswordAction,
  destroyPageAction,
} from './actions';

export default function createResetPasswordContainer(PageComponent, options = {}) {
  const mapStateToProps = createStructuredSelector({
    ResetPasswordPage: selectResetPasswordPage,
  });

  const mapDispatchToProps = {
    onSubmitForm: resetPasswordAction,
    onUnMount: destroyPageAction,
  };

  @connect(mapStateToProps, mapDispatchToProps)
  @compose(AuthenticationContext)
  @injectIntl
  class ResetPasswordContainer extends PureComponent {
    static propTypes = {
      intl: PropTypes.object.isRequired,
      ResetPasswordPage: PropTypes.object.isRequired,
      onUnMount: PropTypes.func.isRequired,
    };

    componentWillUnmount() {
      this.props.onUnMount();
    }

    getErrorMessage() {
      let errorMessage = this.props.ResetPasswordPage.errorMessage;

      if (errorMessage) {
        errorMessage = messages[errorMessage];
      }

      return errorMessage;
    }

    render() {
      const { formatMessage } = this.props.intl;
      const pageTitle = formatMessage(messages.pageTitle);
      const pageDescription = formatMessage(messages.pageDescription);

      return (
        <div className={options.className || 'reset-password-page'}>
          {!options.noHelmet &&
            <Helmet
              title={pageTitle}
              meta={[
                { name: 'description', content: pageDescription },
              ]}
            />
          }

          <PageComponent errorMessage={this.getErrorMessage()} {...this.props} />
        </div>
      );
    }
  }

  return ResetPasswordContainer;
}
