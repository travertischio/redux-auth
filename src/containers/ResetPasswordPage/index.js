/*
 *
 * ResetPasswordPage
 *
 */
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { Link } from 'react-router';
import createResetPasswordContainer from './create-container';
import ResetPasswordForm from '../../components/ResetPasswordForm';
import messages from './messages';

class ResetPasswordPage extends PureComponent {
  static propTypes = {
    ResetPasswordPage: PropTypes.object.isRequired,
    match: PropTypes.object.isRequired,
    isAuthenticated: PropTypes.bool.isRequired,
    onSubmitForm: PropTypes.func.isRequired,
    errorMessage: PropTypes.object,
  };

  static defaultProps = {
    errorMessage: null,
  }

  renderSuccessMessage() {
    return (
      <div>
        <h1><FormattedMessage {...messages.successHeader} /></h1>
        <div>
          {this.renderGoNextBtn()}
        </div>
      </div>
    );
  }

  renderGoNextBtn() {
    if (this.props.isAuthenticated) {
      return (
        <Link to="/home">
          <FormattedMessage {...messages.goToHomeBtnLabel} />
        </Link>
      );
    }

    return (
      <Link to="/sign-in">
        <FormattedMessage {...messages.goToSignInBtnLabel} />
      </Link>
    );
  }

  renderFormOrError() {
    const loading = this.props.ResetPasswordPage.loading;
    const errorMessage = this.props.errorMessage;

    return (
      <div>
        <h1><FormattedMessage {...messages.header} /></h1>

        { loading && <div>Processing... Please wait.</div> }
        { errorMessage && <p><FormattedMessage {...errorMessage} /></p> }

        {this.renderFormOrErrorInner()}
      </div>
    );
  }

  renderFormOrErrorInner() {
    const invalidToken = this.props.ResetPasswordPage.invalidToken;

    if (invalidToken) {
      return this.renderGoToRequestPasswordPageLink();
    }

    return this.renderForm();
  }

  renderGoToRequestPasswordPageLink() {
    return (
      <Link to="/request-password-reset">
        <FormattedMessage {...messages.requestNewPasswordBtnLabel} />
      </Link>
    );
  }

  renderForm() {
    const {
      onSubmitForm,
      match,
    } = this.props;

    const initialValues = {
      token: match.params.resetPasswordToken,
    };

    return (
      <div>
        <ResetPasswordForm
          onSubmit={onSubmitForm}
          initialValues={initialValues}
        />
      </div>
    );
  }

  render() {
    const success = this.props.ResetPasswordPage.success;

    if (success) {
      return this.renderSuccessMessage();
    }

    return this.renderFormOrError();
  }
}

export default createResetPasswordContainer(ResetPasswordPage);
