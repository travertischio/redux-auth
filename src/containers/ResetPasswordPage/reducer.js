/*
 *
 * ResetPasswordPage reducer
 *
 */

import { fromJS } from 'immutable';
import _get from 'lodash/get';
import _camelCase from 'lodash/camelCase';
import _upperFirst from 'lodash/upperFirst';
import {
  RESET_PASSWORD_ACTION,
  RESET_PASSWORD_SUCCESS_ACTION,
  RESET_PASSWORD_FAILED_ACTION,
  DESTROY_PAGE_ACTION,
} from './constants';

const initialState = fromJS({});

function resetPasswordPageReducer(state = initialState, action) {
  switch (action.type) {
    case RESET_PASSWORD_ACTION:
      return onResetPasswordAction(state);
    case RESET_PASSWORD_SUCCESS_ACTION:
      return onResetPasswordSuccessAction(state);
    case RESET_PASSWORD_FAILED_ACTION:
      return onResetPasswordFailedAction(state, action.payload);
    case DESTROY_PAGE_ACTION:
      return onDestroyPageAction(state);
    default:
      return state;
  }
}

function onResetPasswordAction(state) {
  return state.merge({
    loading: true,
    errorMessage: null,
    success: false,
  });
}

function onResetPasswordSuccessAction(state) {
  return state.merge({
    loading: false,
    errorMessage: null,
    success: true,
  });
}

function onResetPasswordFailedAction(state, rejection) {
  const errors = rejection.response.data;
  const newPasswordError = _get(errors, ['new_password', 0]);
  let errorMessage = _get(errors, ['non_field_errors', 0]);
  let invalidToken = true;

  if (newPasswordError === 'WRONG_PASSWORD_8_CHARS_AND_NUMBER') {
    errorMessage = newPasswordError;
    invalidToken = false;
  }

  errorMessage = _camelCase(errorMessage);
  errorMessage = _upperFirst(errorMessage);
  errorMessage = `serverError${errorMessage}`;

  return state.merge({
    loading: false,
    success: false,
    errorMessage,
    invalidToken,
  });
}

function onDestroyPageAction() {
  return fromJS({});
}

export default resetPasswordPageReducer;
