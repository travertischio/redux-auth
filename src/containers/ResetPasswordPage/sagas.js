import {
  takeLatest,
  takeEvery,
  take,
  cancel,
  put,
  call,
} from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import { resetPassword as resetPasswordApiCall } from '../../api';
import { handleAuthenticationSaga } from '../AuthenticationProvider/sagas';
import {
  resetPasswordSuccessAction,
  resetPasswordFailedAction,
} from './actions';
import {
  RESET_PASSWORD_ACTION,
  RESET_PASSWORD_SUCCESS_ACTION,
} from './constants';

export function* defaultSaga() {
  const resetPasswordActionWatcher = yield takeLatest(RESET_PASSWORD_ACTION, resetPasswordSaga);
  const resetPasswordSuccessActionWatcher = yield takeEvery(RESET_PASSWORD_SUCCESS_ACTION, handleAuthenticationSaga);

  yield take(LOCATION_CHANGE);
  yield cancel(resetPasswordActionWatcher);
  yield cancel(resetPasswordSuccessActionWatcher);
}

export function* resetPasswordSaga(action) {
  try {
    const response = yield call(resetPasswordApiCall, action.payload);
    yield put(resetPasswordSuccessAction(response));
  } catch (error) {
    yield put(resetPasswordFailedAction(error));
  }
}

export default [
  defaultSaga,
];
