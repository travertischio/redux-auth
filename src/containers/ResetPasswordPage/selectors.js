import { createSelector } from 'reselect';

const selectResetPasswordPageDomain = (state) => state.get('resetPasswordPage');

const selectResetPasswordPage = createSelector(
  selectResetPasswordPageDomain,
  (substate) => substate.toJS()
);

export default selectResetPasswordPage;
export {
  selectResetPasswordPageDomain,
};
