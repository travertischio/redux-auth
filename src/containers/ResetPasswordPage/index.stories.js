import React from 'react';
import { storiesOf } from '@kadira/storybook';
import ResetPasswordPage from '.';

const resetPasswordToken = 'MQ-NG5uLTI5Yjc3ZGZjZTM3NTQ2YTI5YTY3';

storiesOf('Reset Password Page', module)
  .add('with valid token', () => (
    <ResetPasswordPage
      isAuthenticated
      routeParams={{ resetPasswordToken }}
    />
  ))
  .add('with invalid token', () => (
    <ResetPasswordPage
      isAuthenticated={false}
      routeParams={{ resetPasswordToken: 'NOT_VALID_TOKEN' }}
    />
  ));
