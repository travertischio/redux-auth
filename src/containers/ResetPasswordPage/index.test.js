import React from 'react';
import {
  getStoreWithInitialState,
  findActionByType,
  mountWithIntl,
  createComponentWithIntl,
} from 'react-unit-testing-utils';
import ResetPasswordPage from './index';
import { destroyPageAction } from './actions';
import {
  RESET_PASSWORD_ACTION,
  DESTROY_PAGE_ACTION,
} from './constants';

describe('<ResetPasswordPage />', () => {
  it('should dispatch action RESET_PASSWORD_ACTION when fill and submit the form', () => {
    const store = getStoreWithInitialState({
      auth: {
        isAuthenticated: false,
      },
      resetPasswordPage: {},
    });

    const routeParams = { resetPasswordToken: 'XYZ123' };
    const wrapper = mountWithIntl(<ResetPasswordPage routeParams={routeParams} />, {
      store,
    });

    const newPassword = 'pass123';

    wrapper.find('input[name="new_password"]').simulate('change', { target: { value: newPassword } });
    wrapper.find('input[name="re_new_password"]').simulate('change', { target: { value: newPassword } });
    wrapper.find('form').simulate('submit');

    const recivedAction = findActionByType(store, RESET_PASSWORD_ACTION);
    expect(recivedAction.type).toEqual(RESET_PASSWORD_ACTION);
  });

  it('should dispatch action DESTROY_PAGE_ACTION when the component unmount', () => {
    const initialState = {
      resetPasswordPage: {},
    };
    const { component, store } = renderResetPasswordPageWithState(initialState);
    component.unmount();
    const recivedAction = findActionByType(store, DESTROY_PAGE_ACTION);
    expect(recivedAction).toEqual(destroyPageAction());
  });

  it('should render ResetPasswordPage with the reset password form', () => {
    const initialState = {
      resetPasswordPage: {},
    };
    expectComponentWithStateToMatchSnapshot(initialState);
  });

  it('should render ResetPasswordPage with loading indicator when loading is in the props', () => {
    const initialState = {
      resetPasswordPage: {
        loading: true,
      },
    };
    expectComponentWithStateToMatchSnapshot(initialState);
  });

  it('should render ResetPasswordPage with success message', () => {
    const initialState = {
      auth: {
        isAuthenticated: false,
      },
      resetPasswordPage: {
        success: true,
      },
    };
    expectComponentWithStateToMatchSnapshot(initialState);
  });

  it('should render ResetPasswordPage with success message', () => {
    const initialState = {
      auth: {
        isAuthenticated: true,
      },
      resetPasswordPage: {
        success: true,
      },
    };
    expectComponentWithStateToMatchSnapshot(initialState);
  });

  it('should render ResetPasswordPage with invalid token error message', () => {
    const initialState = {
      resetPasswordPage: {
        invalidToken: true,
      },
    };
    expectComponentWithStateToMatchSnapshot(initialState);
  });

  it('should render ResetPasswordPage with invalid token error message', () => {
    const initialState = {
      resetPasswordPage: {
        invalidToken: true,
      },
    };
    expectComponentWithStateToMatchSnapshot(initialState);
  });

  it('should render ResetPasswordPage with error message when errorMessage is in the props', () => {
    const initialState = {
      resetPasswordPage: {
        errorMessage: 'serverErrorInvalid',
      },
    };
    expectComponentWithStateToMatchSnapshot(initialState);
  });
});

function expectComponentWithStateToMatchSnapshot(initialState) {
  const { component } = renderResetPasswordPageWithState(initialState);
  expect(component.toJSON()).toMatchSnapshot();
}

function renderResetPasswordPageWithState(initialState) {
  const routeParams = { resetPasswordToken: 'XYZ123' };
  return createComponentWithIntl(<ResetPasswordPage routeParams={routeParams} />, initialState);
}
