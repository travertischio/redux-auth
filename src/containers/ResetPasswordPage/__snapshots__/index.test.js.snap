// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetPasswordPage /> should render ResetPasswordPage with error message when errorMessage is in the props 1`] = `
<div>
  <div>
    <h1>
      <span>
        Choose a new password
      </span>
    </h1>
    <p>
      <span>
        Oops, that token has expired. Please click the button below to request new password.
      </span>
    </p>
    <div>
      <form
        noValidate={true}
        onSubmit={[Function]}
      >
        <div>
          <label
            htmlFor="new_password"
          >
            New password
          </label>
          <div
            className=""
          >
            <input
              id="new_password"
              name="new_password"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              placeholder="Enter new password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div>
          <label
            htmlFor="re_new_password"
          >
            Retype new password
          </label>
          <div
            className=""
          >
            <input
              id="re_new_password"
              name="re_new_password"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              placeholder="Retype new password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div>
          <button
            disabled={true}
            type="submit"
          >
            Save
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`<ResetPasswordPage /> should render ResetPasswordPage with invalid token error message 1`] = `
<div>
  <div>
    <h1>
      <span>
        Choose a new password
      </span>
    </h1>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <span>
        Request new password
      </span>
    </a>
  </div>
</div>
`;

exports[`<ResetPasswordPage /> should render ResetPasswordPage with invalid token error message 2`] = `
<div>
  <div>
    <h1>
      <span>
        Choose a new password
      </span>
    </h1>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <span>
        Request new password
      </span>
    </a>
  </div>
</div>
`;

exports[`<ResetPasswordPage /> should render ResetPasswordPage with loading indicator when loading is in the props 1`] = `
<div>
  <div>
    <h1>
      <span>
        Choose a new password
      </span>
    </h1>
    <div>
      Processing... Please wait.
    </div>
    <div>
      <form
        noValidate={true}
        onSubmit={[Function]}
      >
        <div>
          <label
            htmlFor="new_password"
          >
            New password
          </label>
          <div
            className=""
          >
            <input
              id="new_password"
              name="new_password"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              placeholder="Enter new password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div>
          <label
            htmlFor="re_new_password"
          >
            Retype new password
          </label>
          <div
            className=""
          >
            <input
              id="re_new_password"
              name="re_new_password"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              placeholder="Retype new password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div>
          <button
            disabled={true}
            type="submit"
          >
            Save
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`<ResetPasswordPage /> should render ResetPasswordPage with success message 1`] = `
<div>
  <div>
    <h1>
      <span>
        Your password has been changed successfully.
      </span>
    </h1>
    <div>
      <a
        onClick={[Function]}
        style={Object {}}
      >
        <span>
          Go to sign in page.
        </span>
      </a>
    </div>
  </div>
</div>
`;

exports[`<ResetPasswordPage /> should render ResetPasswordPage with the reset password form 1`] = `
<div>
  <div>
    <h1>
      <span>
        Choose a new password
      </span>
    </h1>
    <div>
      <form
        noValidate={true}
        onSubmit={[Function]}
      >
        <div>
          <label
            htmlFor="new_password"
          >
            New password
          </label>
          <div
            className=""
          >
            <input
              id="new_password"
              name="new_password"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              placeholder="Enter new password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div>
          <label
            htmlFor="re_new_password"
          >
            Retype new password
          </label>
          <div
            className=""
          >
            <input
              id="re_new_password"
              name="re_new_password"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              placeholder="Retype new password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div>
          <button
            disabled={true}
            type="submit"
          >
            Save
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
